### PR TITLE
Offline staking implementation

### PR DESCRIPTION
**Pull request by timemarkovqtum**
New command line parameters:
`-offlinestakingheight, -delegationsaddress, -stakingminutxovalue, -stakingminfee, -superstaking`

New RPC calls:
`getdelegationinfoforaddress, getdelegationsforstaker, splitutxosforaddress, removedelegationforaddress, setdelegateforaddress`

The contract for delegation is not added into the EVM configuration because it is required modifications in the genesis bock, parameter `genesis.hashStateRoot` need to be modified in `chainparams.cpp`. The hash state root is a hash from all of the contracts bytecode into the EVM configuration, adding more bytecode will change that hash, no option for start block for compiled bytecode contracts. The precompiled contracts doesn't contain bytecode so they are not included into computation of the hash state root.

Activation from `chainparams.cpp`. Set the offline staking start block and the delegation contract address (create the delegation contract in a block before the offline stake height) for the specific network configuration:
`consensus.nOfflineStakeHeight`
`consensus.delegationsAddress`

For regtest mode those two parameters can be overwritten from command line `-offlinestakingheight` and `-delegationsaddress`.

Regtest command line:
`qtum-qt -regtest -offlinestakingheight=BLOCK_HEIGHT -delegationsaddress=CONTRACT_ADDRESS`

Regtest command line for super staker:
`qtum-qt -regtest -offlinestakingheight=BLOCK_HEIGHT -delegationsaddress=CONTRACT_ADDRESS -superstaking`

### CODE REVIEW

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -1078,6 +1164,210 @@ static UniValue sendtocontract(const JSONRPCRequest& request){
...
+                    "[\n"
+                    "  {\n"
+                    "    \"txid\" : (string) The transaction id.\n"
+                    "    \"sender\" : (string) " + CURRENCY_UNIT + " address of the sender.\n"
+                    "    \"hash160\" : (string) ripemd-160 hash of the sender.\n"
+                    "  }\n"
+                    "]\n"
+                    },
+                    RPCExamples{
+                    HelpExampleCli("setdelegateforaddress", " \"QM72Sfpbz1BPpXFHz9m3CdqATR44Jvaydd\" 10 \"QX1GkJdye9WoUnrE2v6ZQhQ72EUVDtGXQX 6000000\" "+FormatMoney(minGasPrice))
```
Comment (oyyq99999) [inline src/wallet/rpcwallet.cpp]: Typo here
```suggestion
                    HelpExampleCli("setdelegateforaddress", " \"QM72Sfpbz1BPpXFHz9m3CdqATR44Jvaydd\" 10 \"QX1GkJdye9WoUnrE2v6ZQhQ72EUVDtGXQX\" 6000000 "+FormatMoney(minGasPrice))
```

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -2681,16 +2681,50 @@ bool CWallet::HaveAvailableCoinsForStaking() const
     return vCoins.size() > 0;
 }
 
-void CWallet::AvailableDelegateCoinsForStaking(interfaces::Chain::Lock& locked_chain, std::vector<COutput>& vDelegateCoins) const
+bool heightUtxoSort(std::pair<CAddressUnspentKey, CAddressUnspentValue> a,
+                std::pair<CAddressUnspentKey, CAddressUnspentValue> b) {
+    return a.second.blockHeight < b.second.blockHeight;
+}
```
Comment (xuanyan0x7c7) [inline src/wallet/wallet.cpp]: Should pass parameters by const reference instead of by value

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -3797,7 +3797,187 @@ bool CWallet::CreateCoinStakeFromMine(interfaces::Chain::Lock& locked_chain, con
...
+            // Split the reward when mpos is used
+            nRewardPiece = nTotalReward / consensusParams.nMPoSRewardRecipients;
+            int64_t nRewardDelegate = nRewardPiece + nTotalReward % consensusParams.nMPoSRewardRecipients;
+            nRewardStaker = nRewardDelegate * 100 / delegation.fee;
+            nCredit += nRewardDelegate - nRewardStaker;
+        }
+    }
+
+    // Set output amount
+    txNew.vout[1].nValue = nRewardStaker;
```
Comment (xuanyan0x7c7) [inline src/wallet/wallet.cpp]: Should prevent generating zero value utxo when `delegation.fee=0`

File: src/pos.cpp [hunk 4]
```diff
@@ -326,6 +326,15 @@ unsigned int GetStakeSplitOutputs() { return 2; }
 
 int64_t GetStakeSplitThreshold() { return GetStakeSplitOutputs() * GetStakeCombineThreshold(); }
 
+bool SplitOfflineStakeReward(const int64_t& nReward, const uint8_t& fee, int64_t& nRewardOffline, int64_t& nRewardStaker)
+{
+    if(fee > 100) return false;
+    double percentage = fee / 100.0;
+    nRewardStaker = percentage * nReward;
+    nRewardOffline = nReward - nRewardStaker;
```
Comment (xuanyan0x7c7) [inline src/pos.cpp]: Float point calculations are not necessary, `nReward * fee / 100` is ok

File: src/validation.cpp [hunk 5]
```diff
@@ -4738,6 +4748,46 @@ bool GetBlockPublicKey(const CBlock& block, std::vector<unsigned char>& vchPubKe
...
+    if(nValueCoin <= 0)
+        return false;
+
+    CAmount nValueStaker = block.vtx[1]->vout[0].nValue;
+    CAmount nValueDelegate = block.vtx[1]->vout[1].nValue;
+    CAmount nReward = nValueStaker + nValueDelegate - nValueCoin;
+    if(nReward < 0)
+        return false;
+
+    fee = std::lround(nValueStaker * 100.0 / nReward);
```
Comment (xuanyan0x7c7) [inline src/validation.cpp]: `nValueStaker = nReward * fee / 100`, so `fee = std::ceill(nValueStaker * 100.0 / nReward)`

File: src/miner.cpp [hunk 6]
```diff
@@ -917,12 +917,14 @@ void ThreadStakeMiner(CWallet *pwallet, CConnman* connman)
         CAmount nTargetValue = nBalance - pwallet->m_reserve_balance;
         CAmount nValueIn = 0;
         std::set<std::pair<const CWalletTx*,unsigned int> > setCoins;
+        std::set<std::pair<const CWalletTx*,unsigned int> > setDelegateCoins;
         int64_t start = GetAdjustedTime();
         {
             auto locked_chain = pwallet->chain().lock();
             pwallet->SelectCoinsForStaking(*locked_chain, nTargetValue, setCoins, nValueIn);
```
Comment (xuanyan0x7c7) [inline src/miner.cpp]: Addresses in this wallet which are delegated to others should not be included in `setCoins`

File: src/validation.cpp (hunk 5 above)
Comment (xuanyan0x7c7) [reply src/validation.cpp]: And it may produce a `0.0/0.0` when block subsidy reduces to zero

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (timemarkovqtum) [reply src/wallet/rpcwallet.cpp]: Fixed

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (timemarkovqtum) [reply src/wallet/wallet.cpp]: Fixed

File: src/pos.cpp (hunk 4 above)
Comment (timemarkovqtum) [reply src/pos.cpp]: Fixed

File: src/validation.cpp (hunk 5 above)
Comment (timemarkovqtum) [reply src/validation.cpp]: Fixed

File: src/wallet/wallet.cpp (hunk 3 above)
Comment (timemarkovqtum) [reply src/wallet/wallet.cpp]: 0 and 100 fees are allowed in the delegation contract, so if one of those fees are used then one output will be 0 and the other output will contain the full reward. @qtum-neil Does modification is required?

File: src/miner.cpp (hunk 6 above)
Comment (timemarkovqtum) [reply src/miner.cpp]: This is related or present in the hybrid staking issues described by caspal. @qtum-neil Does modification is required?

**Pull request merged**